File: Arkanoid.cs
// Arkanoid.cs
using System;
using System.Collections.Generic;
using System.Threading;

class Arkanoid
{
    const int W = 40, H = 20;
    static int paddleX = W / 2 - 4;
    const int paddleWidth = 8;
    static int ballX = paddleX + paddleWidth / 2;
    static int ballY = H - 3;
    static int ballDX = 1;
    static int ballDY = -1;
    static bool ballLaunched = false;
    static int lives = 3;
    static int score = 0;
    static int level = 1;
    static List<(int x, int y, bool alive)> bricks = new List<(int, int, bool)>();
    static bool gameOver = false;
    static bool running = true;

    static void InitBricks()
    {
        bricks.Clear();
        int rows = Math.Min(3 + level, 6);
        int cols = W / 3;
        for (int y = 0; y < rows; y++)
            for (int x = 1; x < cols; x++)
                bricks.Add((x * 3, y + 2, true));
    }

    static void DrawBorder()
    {
        for (int x = 0; x < W; x++)
        {
            Console.SetCursorPosition(x, 0);
            Console.Write('#');
            Console.SetCursorPosition(x, H - 1);
            Console.Write('#');
        }
        for (int y = 0; y < H; y++)
        {
            Console.SetCursorPosition(0, y);
            Console.Write('#');
            Console.SetCursorPosition(W - 1, y);
            Console.Write('#');
        }
    }

    static void DrawPaddle()
    {
        for (int x = 0; x < paddleWidth; x++)
        {
            Console.SetCursorPosition(paddleX + x, H - 2);
            Console.Write('=');
        }
    }

    static void DrawBall()
    {
        if (ballLaunched || !gameOver)
        {
            Console.SetCursorPosition(ballX, ballY);
            Console.Write('O');
        }
    }

    static void DrawBricks()
    {
        foreach (var b in bricks)
            if (b.alive)
            {
                Console.SetCursorPosition(b.x, b.y);
                Console.Write('█');
            }
    }

    static void DrawInfo()
    {
        Console.SetCursorPosition(W + 2, 0);
        Console.Write($"Score: {score}");
        Console.SetCursorPosition(W + 2, 1);
        Console.Write($"Lives: {lives}");
        Console.SetCursorPosition(W + 2, 2);
        Console.Write($"Level: {level}");
        if (gameOver)
        {
            Console.SetCursorPosition(W / 2 - 5, H / 2);
            Console.Write("GAME OVER! Press R");
        }
        else if (!ballLaunched)
        {
            Console.SetCursorPosition(W / 2 - 5, H / 2);
            Console.Write("Press Space to launch");
        }
    }

    static void UpdateBall()
    {
        if (!ballLaunched || gameOver) return;
        ballX += ballDX;
        ballY += ballDY;
        // walls
        if (ballX <= 1 || ballX >= W - 2) ballDX *= -1;
        if (ballY <= 1) ballDY *= -1;
        // paddle
        if (ballY == H - 2 && ballX >= paddleX && ballX < paddleX + paddleWidth)
        {
            ballDY *= -1;
            int hitPos = ballX - (paddleX + paddleWidth / 2);
            if (hitPos != 0)
                ballDX = hitPos > 0 ? 1 : -1;
        }
        // bricks
        for (int i = 0; i < bricks.Count; i++)
        {
            if (bricks[i].alive && ballY == bricks[i].y && ballX == bricks[i].x)
            {
                var b = bricks[i];
                b.alive = false;
                bricks[i] = b;
                score += 10;
                ballDY *= -1;
                break;
            }
        }
        // bottom
        if (ballY >= H - 1)
        {
            lives--;
            ballLaunched = false;
            ballX = paddleX + paddleWidth / 2;
            ballY = H - 3;
            ballDX = new Random().Next(2) == 0 ? 1 : -1;
            ballDY = -1;
            if (lives == 0) gameOver = true;
        }
        // level complete
        bool allDead = true;
        foreach (var b in bricks) if (b.alive) { allDead = false; break; }
        if (allDead)
        {
            level++;
            InitBricks();
            ballLaunched = false;
            ballX = paddleX + paddleWidth / 2;
            ballY = H - 3;
            ballDX = new Random().Next(2) == 0 ? 1 : -1;
            ballDY = -1;
        }
    }

    static void MovePaddle(int dx)
    {
        if (gameOver) return;
        int newX = paddleX + dx;
        if (newX >= 1 && newX + paddleWidth <= W - 1)
        {
            paddleX = newX;
            if (!ballLaunched)
                ballX = paddleX + paddleWidth / 2;
        }
    }

    static void LaunchBall()
    {
        if (gameOver) return;
        if (!ballLaunched) ballLaunched = true;
    }

    static void Restart()
    {
        paddleX = W / 2 - 4;
        ballX = paddleX + paddleWidth / 2;
        ballY = H - 3;
        ballDX = 1;
        ballDY = -1;
        ballLaunched = false;
        lives = 3;
        score = 0;
        level = 1;
        gameOver = false;
        InitBricks();
    }

    static void InputLoop()
    {
        while (running)
        {
            var key = Console.ReadKey(true);
            switch (key.Key)
            {
                case ConsoleKey.R:
                    if (gameOver) Restart();
                    break;
                case ConsoleKey.Spacebar:
                    LaunchBall();
                    break;
                case ConsoleKey.LeftArrow:
                case ConsoleKey.A:
                    MovePaddle(-2);
                    break;
                case ConsoleKey.RightArrow:
                case ConsoleKey.D:
                    MovePaddle(2);
                    break;
                case ConsoleKey.Q:
                    running = false;
                    return;
            }
        }
    }

    static void Main()
    {
        Console.CursorVisible = false;
        InitBricks();
        Thread inputThread = new Thread(InputLoop);
        inputThread.IsBackground = true;
        inputThread.Start();
        while (running)
        {
            UpdateBall();
            Console.Clear();
            DrawBorder();
            DrawPaddle();
            DrawBall();
            DrawBricks();
            DrawInfo();
            Thread.Sleep(50);
        }
    }
}

File: arkanoid.go
// arkanoid.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"time"
)

const (
	width  = 40
	height = 20
)

type Brick struct{ x, y int; alive bool }

var (
	paddleX      = width/2 - 4
	paddleWidth  = 8
	ballX        = paddleX + paddleWidth/2
	ballY        = height - 3
	ballDX       = 1
	ballDY       = -1
	ballLaunched = false
	lives        = 3
	score        = 0
	level        = 1
	bricks       []Brick
	gameOver     = false
	running      = true
)

func clearScreen() {
	cmd := exec.Command("clear")
	cmd.Stdout = os.Stdout
	cmd.Run()
}

func initBricks() {
	bricks = []Brick{}
	rows := 3 + level
	if rows > 6 {
		rows = 6
	}
	cols := width / 3
	for y := 0; y < rows; y++ {
		for x := 1; x < cols; x++ {
			bricks = append(bricks, Brick{x: x * 3, y: y + 2, alive: true})
		}
	}
}

func drawBorder() {
	for x := 0; x < width; x++ {
		fmt.Printf("\033[%d;%dH#", 0, x)
		fmt.Printf("\033[%d;%dH#", height-1, x)
	}
	for y := 0; y < height; y++ {
		fmt.Printf("\033[%d;%dH#", y, 0)
		fmt.Printf("\033[%d;%dH#", y, width-1)
	}
}

func drawPaddle() {
	for x := 0; x < paddleWidth; x++ {
		fmt.Printf("\033[%d;%dH=", height-2, paddleX+x)
	}
}

func drawBall() {
	if ballLaunched || !gameOver {
		fmt.Printf("\033[%d;%dHO", ballY, ballX)
	}
}

func drawBricks() {
	for _, b := range bricks {
		if b.alive {
			fmt.Printf("\033[%d;%dH█", b.y, b.x)
		}
	}
}

func drawInfo() {
	fmt.Printf("\033[%d;%dHScore: %d", 0, width+2, score)
	fmt.Printf("\033[%d;%dHLives: %d", 1, width+2, lives)
	fmt.Printf("\033[%d;%dHLevel: %d", 2, width+2, level)
	if gameOver {
		fmt.Printf("\033[%d;%dHGAME OVER! Press R", height/2, width/2-5)
	} else if !ballLaunched {
		fmt.Printf("\033[%d;%dHPress Space to launch", height/2, width/2-5)
	}
}

func updateBall() {
	if !ballLaunched || gameOver {
		return
	}
	ballX += ballDX
	ballY += ballDY
	// walls
	if ballX <= 1 || ballX >= width-2 {
		ballDX *= -1
	}
	if ballY <= 1 {
		ballDY *= -1
	}
	// paddle
	if ballY == height-2 && ballX >= paddleX && ballX < paddleX+paddleWidth {
		ballDY *= -1
		hitPos := ballX - (paddleX + paddleWidth/2)
		if hitPos != 0 {
			if hitPos > 0 {
				ballDX = 1
			} else {
				ballDX = -1
			}
		}
	}
	// bricks
	for i := range bricks {
		if bricks[i].alive && ballY == bricks[i].y && ballX == bricks[i].x {
			bricks[i].alive = false
			score += 10
			ballDY *= -1
			break
		}
	}
	// bottom
	if ballY >= height-1 {
		lives--
		ballLaunched = false
		ballX = paddleX + paddleWidth/2
		ballY = height - 3
		ballDX = 1
		if rand.Intn(2) == 0 {
			ballDX = -1
		}
		ballDY = -1
		if lives == 0 {
			gameOver = true
		}
	}
	// level complete
	allDead := true
	for _, b := range bricks {
		if b.alive {
			allDead = false
			break
		}
	}
	if allDead {
		level++
		initBricks()
		ballLaunched = false
		ballX = paddleX + paddleWidth/2
		ballY = height - 3
		ballDX = 1
		if rand.Intn(2) == 0 {
			ballDX = -1
		}
		ballDY = -1
	}
}

func movePaddle(dx int) {
	if gameOver {
		return
	}
	newX := paddleX + dx
	if newX >= 1 && newX+paddleWidth <= width-1 {
		paddleX = newX
		if !ballLaunched {
			ballX = paddleX + paddleWidth/2
		}
	}
}

func launchBall() {
	if gameOver {
		return
	}
	if !ballLaunched {
		ballLaunched = true
	}
}

func restart() {
	paddleX = width/2 - 4
	ballX = paddleX + paddleWidth/2
	ballY = height - 3
	ballDX = 1
	ballDY = -1
	ballLaunched = false
	lives = 3
	score = 0
	level = 1
	gameOver = false
	initBricks()
}

func render() {
	clearScreen()
	drawBorder()
	drawPaddle()
	drawBall()
	drawBricks()
	drawInfo()
}

func readInput() {
	go func() {
		for running {
			var b [1]byte
			os.Stdin.Read(b[:])
			switch b[0] {
			case 'q', 'Q':
				running = false
				os.Exit(0)
			case 'r', 'R':
				if gameOver {
					restart()
				}
			case ' ':
				launchBall()
			case 'a', 'A':
				movePaddle(-2)
			case 'd', 'D':
				movePaddle(2)
			}
		}
	}()
}

func main() {
	rand.Seed(time.Now().UnixNano())
	fmt.Print("\033[?25l") // hide cursor
	initBricks()
	readInput()
	for running {
		updateBall()
		render()
		time.Sleep(50 * time.Millisecond)
	}
}
